### PR TITLE
Add rollback proposal

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
There is no way to tolerate corrupted slots after we've frozen/voted for a bank

#### Summary of Changes
Introduce a proposal for a rollback mechanism

Fixes #

### CODE REVIEW

File: book/src/rollback.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Rollback
+
+Rollback is necessary when leader misbehavior has the potential to be discovered after validators have already replayed a slot to completion.
+
+# Scenarios where Rollback is Necessary:
+
+1) Leader emits different a different set of blobs for the same slot.
```
Comment (sagar-solana) [inline book/src/rollback.md]: Different a different

**Comment by aeyakovenko** [comment]
@carllin if a block is correct, it shouldn't be marked as "dead".  Blocktree should slash the leader which transmits a duplicate block, but it needs to handle storing both, since the network will pick one as the correct one.  As long as the block is still consistent, the network can produce a valid fork from that block.

**Comment by carllin** [comment]
@anatoly how will we distinguish between different transmissions for the same slot for protocols like repair and voting?

**Comment by aeyakovenko** [comment]
@carllin repair the one you picked, and drop the other one.  until its voted on, just repair what is there and the block will fail.

**Comment by sagar-solana** [comment]
@aeyakovenko, which is the "one you picked"? What if you "picked" a different one from everyone else? Also what do you mean by "repair what is there"? I think what Carl is going for is a widespread (gossip) message to get everyone else to deny the block.

**Comment by aeyakovenko** [comment]
@sagar-solana 
@carllin 

A leader proposes 2 blocks.  Some of the network receives block 1, some block 2, some a mix.

1. mix fails, since it just doesn't work
2. if block 1 or block 2 has a supermajority, it can't remain in the chain

**Comment by carllin** [comment]
@aeyakovenko so for the validators that get a mix, they just keep trying to repair either block 1 or block 2 
, whichever one they see first?

Let's say a validator A picks block 1, and votes on block 1. 

Then once the network picks supermajority for block 2, how will the validator know to switch over? You need the block2 to replay, but this validator A is committed to block 1.

**Comment by sagar-solana** [comment]
> mix fails, since it just doesn't work

silly question, you said we shouldn't mark slots as "dead". If we don't mark the mixed slot as "dead", how does it fail ?

> if block 1 or block 2 has a supermajority, it can't remain in the chain

How do we remove them? Does this mean both blocks should be rejected, regardless of "supermajority"?

**Comment by aeyakovenko** [comment]
@sagar-solana yea. I see what you mean. If the block is junk, the node should probably try to repair both,  the problem is that slashing may not come fast enough in time.

**Comment by garious** [comment]
@carllin, what's going on with this PR?

**Pull request closed without merging**